Dodanie funkcji convertToObject

JoAnna578 opened the pull request:
Funkcja konwertuje string CSS na obiekt JS.
Użyto metod map, filter i reduce, bez pętli for.

JoAnna578 committed:
Dodanie funkcji convertToObject

Funkcja konwertuje string CSS na obiekt JS.
Użyto metod map, filter i reduce, bez pętli for.

mateacademy-ai-mentor commented on src/convertToObject.js:
```diff
@@ -1,12 +1,19 @@
+  
+  const validRules = rules
+    .map(rule => rule.trim())
+    .filter(rule => rule.length);
```
This only removes empty strings but does not exclude candidates that lack a colon. The description requires: **"Ignores empty or invalid declaration lines that do not contain a valid property-value pair (e.g., standalone semicolons or blank lines)."** You should filter out rules that do not contain a colon (and later also filter out empty keys/values) before splitting. See task description for the specification.

mateacademy-ai-mentor commented on src/convertToObject.js:
```diff
@@ -1,12 +1,19 @@
+    .map(rule => rule.trim())
+    .filter(rule => rule.length);
+  
+  const keyValuePairs = validRules.map(rule => rule.split(':').map(item => item.trim()));
```
You split each rule with `rule.split(':')`. The parsing requirements say: **"For each non-empty candidate that contains a colon, split into property and value at the first colon."** Using `split(':')` can produce arrays with more than two items or leave you with a single-item array when no colon exists. Consider splitting at the first colon (indexOf + slice) or validate `rule.includes(':')` before splitting so you always get a key and a value.

mateacademy-ai-mentor commented on src/convertToObject.js:
```diff
@@ -1,12 +1,19 @@
+  const result = keyValuePairs.reduce((obj, [key, value]) => {
+    obj[key] = value;
+    return obj;
+  }, {});
```
This reducer mutates its accumulator and uses generic names. This violates checklist rules: 

- **"[CODE KNOWLEDGE] - Don't use iteration methods to modify an already existing array/object. (forEach is exception). These methods return a new array, so you should use it."**
- **"[CODE STYLE] - Avoid variable namings like `array` , `object` `arr`, `obj`, `result`. It will decrease code readability"**

Instead of `obj[key] = value` inside reduce, either return a new object each iteration (e.g. `{ ...acc, [key]: value }`) or use `forEach` with a clearly named accumulator object (e.g., `stylesMap`) if you intend to mutate. Rename `obj` and `result` to more descriptive identifiers.

mateacademy-ai-mentor commented on src/convertToObject.js:
```diff
@@ -1,12 +1,19 @@
+  const keyValuePairs = validRules.map(rule => rule.split(':').map(item => item.trim()));
+  
+  const result = keyValuePairs.reduce((obj, [key, value]) => {
+    obj[key] = value;
```
As written this line may assign `undefined` values or empty-string keys when a rule lacks a colon or the value is missing. The task requires: **"Trim whitespace around property names and values."** and **"Exclude entries with empty property names after trimming, or with no value part."** Ensure you validate that `key` is non-empty and `value` is present before adding to the output object.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

I am requesting changes because the implementation does not fully meet the parsing requirements and violates explicit checklist items. Specifically, in src/convertToObject.js: (1) you do not filter out declaration candidates that lack a colon before splitting, which can create empty keys or undefined values; ensure you ignore any rule without `:` and only split at the first colon (e.g., via `indexOf`/`slice`). (2) After splitting, you need to trim both sides and skip entries where the property name is empty or the value part is missing; this is required to ignore invalid lines and avoid incorrect object entries. (3) The reducer currently mutates its accumulator (`obj[key] = value`) and uses generic names (`obj`, `result`), which conflicts with the checklist; either return a new object each reduce iteration or switch to `forEach` with a clearly named accumulator like `stylesMap`. 

Minor improvements: prefer descriptive names for intermediate arrays (e.g., `validDeclarations`, `pairs`) and consider guarding for multiple colons by splitting at the first colon to keep the remainder as the value. You’ve got a solid structure with split/map/filter; with these adjustments, your function will be robust and compliant. 

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

JoAnna578 committed:
Poprawa stylu funkcji convertToObject zgodnie z Prettierem

- Usunięto przecinki na końcu obiektu
- Poprawiono wcięcia w łańcuchu metod map/filter/reduce
- Funkcja działa zgodnie z checklistą

JoAnna578 committed:
Poprawka funkcji convertToObject – spełnia wymagania checklisty

- Ignoruje linie CSS bez dwukropka
- Splituje tylko przy pierwszym dwukropku
- Odrzuca puste klucze i wartości
- Tworzy nowy obiekt w reduce, nie mutuje istniejącego
- Zmieniono nazwy zmiennych na bardziej opisowe
- Funkcja zgodna z Prettierem i checklistą

JoAnna578 committed:
Finalna poprawka funkcji convertToObject – spełnia wymagania checklisty

- Ignoruje linie bez dwukropka i puste linie
- Splituje tylko przy pierwszym dwukropku
- Odrzuca puste klucze i wartości
- Tworzy nowy obiekt w reduce, nie mutuje istniejącego
- Użyto opisowych nazw zmiennych: declarations, validDeclarations, stylesMap
- Styl zgodny z Prettier i checklistą Mate Academy

JoAnna578 committed:
Finalna poprawka convertToObject – max-len i zgodność z checklistą

- Wszystkie linie ≤ 80 znaków, zgodnie z ESLint/Prettier
- Ignoruje linie CSS bez dwukropka i puste linie
- Splituje tylko przy pierwszym dwukropku
- Odrzuca puste klucze i wartości
- Tworzy nowy obiekt w reduce, nie mutuje istniejącego
- Nazwy zmiennych opisowe: declarations, validDeclarations, stylesMap
- Funkcja w pełni zgodna z checklistą Mate Academy

JoAnna578 committed:
Zamiana exportu funkcji convertToObject na CommonJS dla testów

- Zmieniono `export default convertToObject` na `module.exports = convertToObject`
- Dzięki temu testy z `require('./convertToObject')` działają poprawnie
- Usunięto konflikt między ESM a CommonJS
- Funkcja pozostaje niezmieniona, tylko sposób eksportu został dopasowany do środowiska testowego

mateacademy-ai-mentor reviewed (approved):
### Review Summary

Your solution is approved. The `convertToObject` function correctly parses the CSS-like string, trims whitespace, ignores malformed or empty declarations, preserves values (including vendor prefixes and colons within values), and returns a new object without mutating intermediate data. It meets the task requirements of using array iteration methods and avoiding loops, and the implementation is clear and readable.

Minor, non-blocking suggestions for future refinement:
- Consider a more domain-specific reducer accumulator name (e.g., `stylesObject`) for clarity.
- Be aware that creating a new object on each `reduce` step can be less performant on very large inputs; a single accumulator object via `forEach` could be an optimization if performance becomes a concern.
- Optionally add a guard to handle non-string inputs by returning `{}` for extra robustness.

Nice work keeping the logic clean and robust.

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)
